### PR TITLE
Add simple pass manager for `GpuLower`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
`GpuLower`'s ctor now only does analysis. Users needs to call `GpuLower::run` to run passes. Passes are `std::vector<std::pair<std::string, std::function<std::vector<Expr*>(const std::vector<Expr*>&)>>`, and are exposes to the public as `GpuLower::passes()`. Users can modify it. The main use case for this mechanism is for testing purposes.

One demo is the test added in `test_gpu3.cpp`:

```C++
TEST_F(NVFuserTest, LoweringHook) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(2);
  fusion.addInput(tv0);
  auto tv1 = set(tv0);
  fusion.addOutput(tv1);

  GpuLower gpulw(&fusion);
  bool executed = false;
  gpulw.passes().push_back(
      {"test",
       [&executed](const std::vector<Expr*>& exprs) -> std::vector<Expr*> {
         executed = true;
         return exprs;
       }});
  EXPECT_FALSE(executed);
  gpulw.run();
  EXPECT_TRUE(executed);
}
```

This should be just refactoring, no codegen change.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build --diff

**Comment by zasdfgbnm** [comment]
!build --diff

**Comment by zasdfgbnm** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
You may want to update and rerun `!build --diff` since I just merged #1214 which should squelch all those false positive PTX diffs.

**Comment by zasdfgbnm** [comment]
!build --diff

**Comment by zasdfgbnm** [comment]
ping @naoyam

File: csrc/device_lower/lower2device.h [hunk 1]
```diff
@@ -51,16 +52,17 @@ class GpuLower : public NonCopyable {
  public:
   GpuLower() = delete;
 
+  using Pass = std::pair<
```
Comment (naoyam) [inline csrc/device_lower/lower2device.h:55]: nit: some comment about `Pass` would be great to have, e.g., what the string means, what the function input and output are.

File: csrc/device_lower/lower2device.h [hunk 2]
```diff
@@ -71,6 +73,9 @@ class GpuLower : public NonCopyable {
   //! Query if lowering is in progress
   static bool hasCurrent();
 
+  //! Actually run the lowering
```
Comment (naoyam) [inline csrc/device_lower/lower2device.h:76]:
```suggestion
  //! Actually run the lowering by executing the passes in the order
```

File: csrc/device_lower/lower2device.h [hunk 3]
```diff
@@ -230,6 +243,9 @@ class GpuLower : public NonCopyable {
   // Lowered Kernel IR
   std::unique_ptr<kir::Kernel> kernel_;
 
+  // Passes to lower kernel
```
Comment (naoyam) [inline csrc/device_lower/lower2device.h:246]: Please also mention they are executed in this order

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the refactoring!

File: csrc/device_lower/lower2device.h (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/device_lower/lower2device.h:246]: added

File: csrc/device_lower/lower2device.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/device_lower/lower2device.h]: added

File: csrc/device_lower/lower2device.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/device_lower/lower2device.h:55]: added

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: test/test_gpu3.cpp [hunk 4]
```diff
@@ -8413,8 +8444,10 @@ TEST_F(NVFuserTest, AlignedSyncReduction1_CUDA) {
 
   scheduler_utils::parallelizeAllLike(tv2);
 
-  const std::string kernel_string =
-      codegen::generateCudaKernel(GpuLower(&fusion).kernel());
+  GpuLower gpulw(&fusion);
+  gpulw.run();
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply test/test_gpu3.cpp]: done

**Pull request merged**